src/ipinfo.rs

```diff
@@ -182,88 +186,78 @@ impl IpInfo {
         batch_config: BatchReqOpts,
     ) -> Result<HashMap<String, IpDetails>, IpError> {
-        let mut hits: Vec<IpDetails> = vec![];
-        let mut misses: Vec<&str> = vec![];
-        let mut to_lookup: Vec<&str> = vec![];
-        let mut details_bogon = vec![];
-        let mut results: HashMap<String, IpDetails> = HashMap::new();
+        // Handle the total timeout condition
+        if let Some(total_timeout) = batch_config.timeout_total {
+            match timeout(total_timeout, self._lookup_batch(ips, batch_config))
+                .await
+            {
+                Ok(result) => result,
+                Err(_) => Err(err!(TimeOutError)),
+            }
+        } else {
+            self._lookup_batch(ips, batch_config).await
+        }
+    }
 
-        // Check for cache hits
-        ips.iter()
-            .for_each(|x| match self.cache.get(&cache_key(x)) {
-                Some(detail) => hits.push(detail.clone()),
-                None => misses.push(*x),
-            });
+    // Internal lookup_batch function. This ignores the total timeout condition
+    async fn _lookup_batch(
+        &mut self,
+        ips: &[&str],
+        batch_config: BatchReqOpts,
+    ) -> Result<HashMap<String, IpDetails>, IpError> {
+        let mut results: HashMap<String, IpDetails> = HashMap::new();
 
-        // check for bogon ip addresses
-        for ip_address in misses.iter() {
-            match is_bogon(ip_address) {
-                true => details_bogon.push(IpDetails {
-                    ip: ip_address.to_string(),
-                    bogon: Some(true),
-                    ..Default::default()
-                }),
-                false => to_lookup.push(ip_address),
+        // Collect a list of ips we need to lookup.
+        // Filters out bogons and cache hits
+        let mut work = vec![];
+        for ip in ips.iter() {
+            if is_bogon(ip) {
+                results.insert(
+                    ip.to_string(),
+                    IpDetails {
+                        ip: ip.to_string(),
+                        bogon: Some(true),
+                        ..Default::default()
+                    },
+                );
+            } else if let Some(detail) = self.cache.get(&cache_key(ip)) {
+                results.insert(ip.to_string(), detail.clone());
+            } else {
+                work.push(*ip);
             }
         }
 
         let client = reqwest::Client::builder()
             .timeout(batch_config.timeout_per_batch)
             .build()?;
-        for i in (0..to_lookup.len()).step_by(batch_config.batch_size as usize)
-        {
-            let mut end = i + batch_config.batch_size as usize;
-            if end > to_lookup.len() {
-                end = to_lookup.len();
-            }
 
-            let lookup_list = &to_lookup[i..end];
+        // Remove duplicates
+        let set: HashSet<&str> = work.iter().cloned().collect();
+        work = set.into_iter().collect();
```

Comment:
```suggestion
        // Remove duplicates
        work.sort();
        work.dedup();
```

I usually prefer this ... it just looks nicer.

Reply:
I like it!